### ISSUE TITLE
Unauthorized Access Tabris 2.0/RAP 3.1

### ISSUE THREAD

**Issue by amedavid** [eclipsesource/tabris#446]
We get this error, if we start your App with Tabris 2.0 and RAP 3.1:

> You are not allowed to connect to the Server

ServerId, Server-Version and Client-Version are the same.

**Comment by ifurnadjiev** [comment]
Hi,
Tabris (RAP) does not manage server access rights in any way. Without 
additional technical details we can't help much. Did you configure some 
access restriction on the app/app server?
Regards,
Ivan

**Comment by ifurnadjiev** [comment]
Hi,
just a hint... please check that the client app based on Tabris 2.0 is 
not running against RAP server 2.3.
HTH,
Ivan

**Comment by amedavid** [comment]
Hi,
we have no access restriction in our App. Only the Tabris required ServerId.

Here the jetty-log:

User-Agent: com.eclipsesource.tabris.android/2.0.0 (com.ametras.moeveapp/2.0.0)

Android-Tabris: 2.0.0
Server: 2.0.0

Regards,

David

**Comment by ifurnadjiev** [comment]
Hi,
User-Agent string is not enough. Can you log the complete request/response, which leads to this error?
Regards,
Ivan

**Comment by amedavid** [comment]
@ifurnadjiev 
[jettylog.txt](https://github.com/eclipsesource/tabris/files/547807/jettylog.txt)

**Comment by amedavid** [comment]
@ifurnadjiev are there any news for us? We have the same Problem with the Tabris iOS-Version.

**Comment by ifurnadjiev** [comment]
Hi,
I don't see any helpful information in the provided Jetty logs. What I still don't understand is where did you get this error "You are not allowed to connect to the Server"? Is it a popup dialog? Is it possible to log the complete HTTP request/response (including headers and body) that lead to this error?
Regards,
Ivan

**Comment by amedavid** [comment]
Hi,
We get this message:
![2016-10-27_10h19_45](https://cloud.githubusercontent.com/assets/19856611/19759823/e745c8c0-9c2e-11e6-8702-4a97414a6362.png)

```
 <meta http-equiv="Content-Type" content="text/html;charset=ISO-8859-1"/>\n
    <title>Error 404 </title>\n
    </head>\n
    <body>\n
    <h2>HTTP ERROR: 404</h2>\n
    <p>Problem accessing /app. Reason:\n
    <pre>    ProxyServlet: /app</pre></p>\n
    <hr /><a href="http://eclipse.org/jetty">Powered by Jetty:// 9.3.9.v20160517</a><hr/>\n
    </body>\n
    </html>\n
```
Regards,
David

**Comment by amedavid** [comment]
@ifurnadjiev do you have enough information?

**Comment by ifurnadjiev** [comment]
The response from the server is 404 - entry point not found. You are sure that the URL you try to connect is correct? Can you access this URL from a browser?

**Comment by amedavid** [comment]
The access by browser works fine. But the problem does occur while accessing the server by app(iOS/Android). This Problem happens since the 2.0 Tabris-Update. With a lower version erverthing works fine. We change nothing on the config. 

For example a Screenshot if we have a wrong URL:
![2016-10-27_13h41_04](https://cloud.githubusercontent.com/assets/19856611/19765953/81caeeae-9c4b-11e6-9d6d-37bfc61621ce.png)

For example a Screenshot if we have a older Server-Version(Version-Check):
![2016-10-27_13h42_35](https://cloud.githubusercontent.com/assets/19856611/19766760/f1625c76-9c4f-11e6-9052-ced9cecfd544.png)

So the question is why we get this message: "You are not allowed to connect to the Server"?

Regards,

David

**Comment by amedavid** [comment]
@ifurnadjiev So we found the error.
If we change the requiredServerId to nil--> everthing works fine. If we put there a String--> Error Message "You are not allowed to connect to the Server". The Server and the Client have the same ServerId.
So is there a change or bug in Tabris 2.0?

Method:
(NSString *)requiredServerId

Regards,

David

**Comment by patrykmol** [comment]
@amedavid This is not a bug. This is has been implemented since 1.3. It is used to prevent random clients from connecting to your server.

**Comment by amedavid** [comment]
Please read my comment again. We use the function since 1.3, but with 2.0 it doesen't work anymore. So when we include a ServerId in the mobile-client we get the error-message.

**Comment by amedavid** [comment]
@ifurnadjiev Can you confirm the problem? 

**For example: We include this in the iOS-App:**
- (NSString *)requiredServerId {
    return @"1234";
}

**Tabris-Server:**
   TabrisClientInstaller.install(application, "1234");

So we get the message: You are not allowed to connect to the Server

We don't change anything on this, but with Tabris 2.0 it doesn't work. What we can do?

**Comment by ifurnadjiev** [comment]
I don't see any server-side Tabris change related to serverId. The serverId is still returned as a response header. @patrykmol, @mpost any idea about it?

**Comment by mpost** [comment]
@ifurnadjiev You are right. There hasn't been a change to the way the id works. Maybe changes related to iOS? @patrykmol?

**Comment by amedavid** [comment]
@mpost same Problem with Android.

**Comment by patrykmol** [comment]
@mpost There where no changes on iOS either afaik. However this problem occurs on both platforms according to this comment https://github.com/eclipsesource/tabris/issues/446#issuecomment-256625123. Therefore I would say something is wrong with server.

**Comment by evolanakis** [comment]
@amedavid Can you verify that the code is actually executed on the server-side? (i.e. by adding a print-out or breakpoint)

**Comment by amedavid** [comment]
@evolanakis Yes i can verify this.

**Comment by ifurnadjiev** [comment]
Maybe the things has changed without notice - in Tabris 1.x (RAP 2.3) the serverId was set as a response header in the first GET request, in Tabris 2.x (RAP 3.0) it's in the first POST request - when the UI session is initialized. @mpost, @patrykmol  can you confirm that you are reading this header from the response of the first GET request?

**Comment by mpost** [comment]
@ifurnadjiev As with tabris 1 the intial get request contains the the requiredServerId

**Comment by ifurnadjiev** [comment]
ok... I think we found it. We have a regression in Tabris 2.0 and will fix it in Tabris 2.0.1.

**Comment by amedavid** [comment]
@ifurnadjiev Ok, perfect. Thanks for the support.

**Comment by amedavid** [comment]
@ifurnadjiev When can we expect with 2.0.1?

**Comment by ifurnadjiev** [comment]
The issue has been fixed on Android, but still needs some work on iOS. I hope Tabris 2.0.1 will be released in a few weeks.

**Comment by ifurnadjiev** [comment]
Fixed in 2.0.1 version.
